aptos-move/aptos-vm-types/src/change_set.rs

```diff
@@ -6,51 +6,166 @@ use aptos_aggregator::delta_change_set::{deserialize, serialize, DeltaChangeSet}
-    write_set::{WriteOp, WriteSet},
+    state_store::state_key::{StateKey, StateKeyInner},
+    transaction::ChangeSet as StorageChangeSet,
+    write_set::{WriteOp, WriteSetMut},
 };
 use move_binary_format::errors::Location;
 use move_core_types::vm_status::{err_msg, StatusCode, VMStatus};
-use std::collections::btree_map::Entry::{Occupied, Vacant};
+use std::collections::{
+    btree_map,
+    btree_map::{
+        Entry,
+        Entry::{Occupied, Vacant},
+    },
+    BTreeMap,
+};
+
+/// Encapsulates any changes to the global state.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct StateChange<K: PartialOrd + Ord, V>(BTreeMap<K, V>);
+
+impl<K, V> StateChange<K, V>
+where
+    K: PartialOrd + Ord,
+{
+    pub fn empty() -> Self {
+        Self(BTreeMap::new())
+    }
+
+    pub fn new(changes: impl IntoIterator<Item = (K, V)>) -> Self {
+        Self(changes.into_iter().collect())
+    }
+
+    #[inline]
+    pub fn insert(&mut self, key: K, value: V) {
+        self.0.insert(key, value);
+    }
+
+    #[inline]
+    pub fn remove(&mut self, key: &K) -> Option<V> {
+        self.0.remove(key)
+    }
+
+    #[inline]
+    pub fn contains(&self, key: &K) -> bool {
+        self.0.contains_key(key)
+    }
+
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    #[inline]
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    #[inline]
+    pub fn get(&self, key: &K) -> Option<&V> {
+        self.0.get(key)
+    }
+
+    #[inline]
+    pub fn get_mut(&mut self, key: &K) -> Option<&mut V> {
+        self.0.get_mut(key)
+    }
+
+    #[inline]
+    pub fn entry(&mut self, key: K) -> Entry<K, V> {
+        self.0.entry(key)
+    }
+
+    #[inline]
+    pub fn iter(&self) -> btree_map::Iter<'_, K, V> {
+        self.0.iter()
+    }
+}
+
+impl<K, V> IntoIterator for StateChange<K, V>
+where
+    K: PartialOrd + Ord,
+{
+    type IntoIter = btree_map::IntoIter<K, V>;
+    type Item = (K, V);
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.0.into_iter()
+    }
+}
 
 /// A change set produced by the VM. Just like VMOutput, this type should
 /// be used inside the VM. For storage backends, use ChangeSet.
 #[derive(Debug, Clone)]
 pub struct VMChangeSet {
-    write_set: WriteSet,
+    // Changes to the data in the global state.
+    resource_write_set: StateChange<StateKey, WriteOp>,
+    // Changes to the code in the global state.
+    module_write_set: StateChange<StateKey, WriteOp>,
+    // Aggregator changes: writes and deltas.
+    // TODO(aggregator): For Aggregator V2, we will not use state keys
+    // (both for writes and deltas)!
+    aggregator_write_set: StateChange<StateKey, WriteOp>,
     delta_change_set: DeltaChangeSet,
+    // Events produced during the execution of a transaction.
     events: Vec<ContractEvent>,
 }
 
+// Useful macro sto squash a pair of write ops and ignore the result if
+// it is a no-op.
+macro_rules! squash_write_ops {
```

Comment:
just wondering, why is this a macro, not a function?

Reply:
Used to avoid `mut`s, etc., since the code is fully shared between two implementations. No strong preference, can make it a function